src/main/java/org/ibp/api/java/impl/middleware/study/validator/ObservationImportRequestValidator.java

```diff
@@ -0,0 +1,193 @@
+        this.errors = new MapBindingResult(new HashMap<>(), ObservationDto.class.getName());
+
+        final List<String> germplasmDbIds = observationDtos.stream().filter(o -> StringUtils.isNotEmpty(o.getGermplasmDbId()))
+                .map(ObservationDto::getGermplasmDbId).collect(Collectors.toList());
+        final GermplasmSearchRequest germplasmSearchRequest = new GermplasmSearchRequest();
+        germplasmSearchRequest.setGermplasmDbIds(germplasmDbIds);
+        final Map<String, GermplasmDTO> germplasmDTOMap = this.germplasmService.searchGermplasmDTO(germplasmSearchRequest, null)
+                .stream().collect(Collectors.toMap(GermplasmDTO::getGermplasmDbId, Function.identity()));
+
+        final List<String> studyDbIds = observationDtos.stream().filter(o -> StringUtils.isNotEmpty(o.getStudyDbId()))
+                .map(ObservationDto::getStudyDbId).collect(Collectors.toList());
+        final StudySearchFilter studySearchFilter = new StudySearchFilter();
+        studySearchFilter.setStudyDbIds(studyDbIds);
+        final Map<String, StudyInstanceDto> studyInstancesMap =
+                this.studyServiceBrapi.getStudyInstances(studySearchFilter, null).stream()
+                        .collect(Collectors.toMap(StudyInstanceDto::getStudyDbId, Function.identity()));
+
+        final List<String> observationUnitDbIds = observationDtos.stream().filter(o -> StringUtils.isNotEmpty(o.getObservationUnitDbId()))
+                .map(ObservationDto::getObservationUnitDbId).collect(Collectors.toList());
+        final ObservationUnitSearchRequestDTO observationUnitSearchRequestDTO = new ObservationUnitSearchRequestDTO();
+        observationUnitSearchRequestDTO.setObservationUnitDbIds(observationUnitDbIds);
+        final Map<String, ObservationUnitDto> observationUnitDtoMap =
+                this.observationUnitService.searchObservationUnits(null, null, observationUnitSearchRequestDTO).stream()
+                        .collect(Collectors.toMap(ObservationUnitDto::getObservationUnitDbId, Function.identity()));
+
+        final List<String> variableIds = observationDtos.stream().filter(o -> StringUtils.isNotEmpty(o.getObservationVariableDbId()))
+                .map(ObservationDto::getObservationVariableDbId).collect(Collectors.toList());
+        final VariableSearchRequestDTO variableSearchRequestDTO = new VariableSearchRequestDTO();
+        variableSearchRequestDTO.setObservationVariableDbIds(variableIds);
+        final Map<String, VariableDTO> variableDTOMap =
+                this.variableServiceBrapi.getObservationVariables(null, variableSearchRequestDTO, null).stream()
+                        .collect(Collectors.toMap(VariableDTO::getObservationVariableDbId, Function.identity()));
+
+        final Map<String, List<String>> studyVariableIdsMap = new HashMap<>();
+
+        Integer index = 1;
+        final Iterator<ObservationDto> iterator = observationDtos.iterator();
+        while (iterator.hasNext()) {
```

Comment:
Instead of iterating and having while loop, we can use:  `observationDtos.removeIf(...)`. Sample usage in `GermplasmImportRequestValidator`